components/psa/services/psa_prot_internal_storage/COMPONENT_PSA_SRV_IMPL/pits_impl.cpp

```diff
@@ -0,0 +1,237 @@
+{
+    // Note: tdb_filename assumed to be of size PSA_ITS_FILENAME_LEN_BYTES and reset with zeros
+
+    uint8_t filename_idx = 0;
+    uint32_t tmp_uid = uid;
+    uint32_t tmp_pid = pid;
+
+    // Iterate on UID; each time convert 6 bits of UID into a character; first iteration must be done
+    do {
+        tdb_filename[filename_idx++] = base64_coding_table[tmp_uid & 0x3F];
+        tmp_uid = tmp_uid >> 6;
+    } while (tmp_uid != 0);
+
+    // Write delimiter
+    tdb_filename[filename_idx++] = '#';
+
+    // Iterate on PID; each time convert 6 bits of PID into a character; first iteration must be done
+    do {
+        tdb_filename[filename_idx++] = base64_coding_table[tmp_pid & 0x3F];
+        tmp_pid = tmp_pid >> 6;
+    } while (tmp_pid != 0);
+}
+
+
+psa_its_status_t psa_its_set_impl(uint32_t pid, uint32_t uid, uint32_t data_length, const void *p_data, psa_its_create_flags_t create_flags)
+{
+    if((create_flags != 0) && (create_flags != PSA_ITS_WRITE_ONCE_FLAG)) {
+        return PSA_ITS_ERROR_FLAGS_NOT_SUPPORTED;
+    }
+
+    KVStore *kvstore = get_kvstore_instance();
+    if (!kvstore) {
+        return PSA_ITS_ERROR_STORAGE_FAILURE;
+    }
+
+    // Generate KVStore key
+    uint8_t kv_key[PSA_ITS_FILENAME_LEN_BYTES] = {'\0'};
+    generate_fn(kv_key, uid, pid);
+
+    uint32_t kv_create_flags = 0;
+    if (create_flags & PSA_ITS_WRITE_ONCE_FLAG) {
+        kv_create_flags = KVStore::WRITE_ONCE_FLAG;
+    }
+
+    int kvstore_status = kvstore->set((const char *)kv_key, p_data, data_length, kv_create_flags);
+
+    psa_its_status_t status = PSA_ITS_SUCCESS;
+    if (kvstore_status != MBED_SUCCESS) {
+        switch(kvstore_status) {
+            case MBED_ERROR_WRITE_PROTECTED:
+                status = PSA_ITS_ERROR_WRITE_ONCE;
+                break;
+            case MBED_ERROR_MEDIA_FULL:
+                status = PSA_ITS_ERROR_INSUFFICIENT_SPACE;
+                break;
+            default:
+                status = PSA_ITS_ERROR_STORAGE_FAILURE;
+        }
+    }
+
+    return status;
+}
+
+psa_its_status_t psa_its_get_impl(uint32_t pid, uint32_t uid, uint32_t data_offset, uint32_t data_length, void *p_data)
+{
+    KVStore *kvstore = get_kvstore_instance();
+    if (!kvstore) {
+        return PSA_ITS_ERROR_STORAGE_FAILURE;
+    }
+
+    // Generate KVStore key
+    uint8_t kv_key[PSA_ITS_FILENAME_LEN_BYTES] = {'\0'};
+    generate_fn(kv_key, uid, pid);
+
+    KVStore::info_t kv_info;
+    int kvstore_status = kvstore->get_info((const char *)kv_key, &kv_info);
+
+    psa_its_status_t status = PSA_ITS_SUCCESS;
+    if (kvstore_status != MBED_SUCCESS) {
+        switch(kvstore_status) {
+            case MBED_ERROR_ITEM_NOT_FOUND:
+                status = PSA_ITS_ERROR_KEY_NOT_FOUND;
+                break;
+            default:
+                status = PSA_ITS_ERROR_STORAGE_FAILURE;
+        }
+    }
+
+    if (kvstore_status == MBED_SUCCESS) {
+        if (data_offset >= kv_info.size) {
```

Comment:
why are we duplicating KVStore checks?
i think `kvstore->get_info` and all these checks are redundant.
@davidsaada can i have your opinion here?

Reply:
why do we need to call get_info API upon get?
the wraparound should be handled by kvstore - if not it should be fixed there.
the `data_offset + data_length > kv_info.size` check can be done just after calling `kvstore->get` API and comparing `actual_size != data_length`.